tools/distrib/python/grpcio_tools/grpc_tools/_protoc_compiler.pyx

```diff
@@ -12,13 +12,114 @@
+        self.line = line
+        self.column = column
+        self.message = message
+
+    def __repr__(self):
+        return "ProtocError(filename=\"{}\", line={}, column={}, message=\"{}\")".format(self.filename, self.line, self.column, self.message)
+
+    def __str__(self):
+        return "{}:{}:{} error: {}".format(self.filename.decode("ascii"), self.line, self.column, self.message.decode("ascii"))
+
+class ProtocWarning(Warning):
+    def __init__(self, filename, line, column, message):
+        self.filename = filename
+        self.line = line
+        self.column = column
+        self.message = message
+
+    def __repr__(self):
+        return "ProtocWarning(filename=\"{}\", line={}, column={}, message=\"{}\")".format(self.filename, self.line, self.column, self.message)
+
+    # TODO: Maybe come up with something better than this
+    __str__ = __repr__
+
+
+class ProtocErrors(Exception):
+    def __init__(self, errors):
+        self._errors = errors
+
+    def errors(self):
+        return self._errors
+
+    def __repr__(self):
+        return "ProtocErrors[{}]".join(repr(err) for err in self._errors)
+
+    def __str__(self):
+        return "\n".join(str(err) for err in self._errors)
+
+cdef _c_protoc_error_to_protoc_error(cProtocError c_protoc_error):
+    return ProtocError(c_protoc_error.filename, c_protoc_error.line, c_protoc_error.column, c_protoc_error.message)
+
+cdef _c_protoc_warning_to_protoc_warning(cProtocWarning c_protoc_warning):
+    return ProtocWarning(c_protoc_warning.filename, c_protoc_warning.line, c_protoc_warning.column, c_protoc_warning.message)
+
+cdef _handle_errors(int rc, vector[cProtocError]* errors, vector[cProtocWarning]* wrnings, bytes protobuf_path):
+  for warning in dereference(wrnings):
+      warnings.warn(_c_protoc_warning_to_protoc_warning(warning))
+  if rc != 0:
+    if dereference(errors).size() != 0:
+       py_errors = [_c_protoc_error_to_protoc_error(c_error) for c_error in dereference(errors)]
+       # TODO: Come up with a good system for printing multiple errors from
+       # protoc.
+       raise ProtocErrors(py_errors)
+    raise Exception("An unknown error occurred while compiling {}".format(protobuf_path))
+
+def get_protos(bytes protobuf_path, list include_paths):
+  cdef vector[string] c_include_paths = include_paths
+  cdef vector[pair[string, string]] files
+  cdef vector[cProtocError] errors
+  # NOTE: Abbreviated name used to avoid shadowing of the module name.
+  cdef vector[cProtocWarning] wrnings
+  rc = protoc_get_protos(protobuf_path, &c_include_paths, &files, &errors, &wrnings)
+  _handle_errors(rc, &errors, &wrnings, protobuf_path)
+  return files
+
+def get_services(bytes protobuf_path, list include_paths):
+  cdef vector[string] c_include_paths = include_paths
+  cdef vector[pair[string, string]] files
+  cdef vector[cProtocError] errors
+  # NOTE: Abbreviated name used to avoid shadowing of the module name.
+  cdef vector[cProtocWarning] wrnings
```

Comment:
warnings? or intended `wrnings`?

Reply:
See the note above. This is unfortunately intentional. If I used the full word, I'd be overwriting the handle to the `warnings` module.